massa-async-pool/src/message.rs

```diff
@@ -0,0 +1,219 @@
+//! Copyright (c) 2022 MASSA LABS <info@massa.net>
+
+//! This file defines the structure representing an entry in the FinalLedger
+
+use massa_models::constants::ADDRESS_SIZE_BYTES;
+use massa_models::{
+    array_from_slice, Address, Amount, DeserializeVarInt, ModelsError, SerializeVarInt, Slot,
+};
+use massa_models::{DeserializeCompact, SerializeCompact};
+use serde::{Deserialize, Serialize};
+
+/// Unique identifier of a message.
+/// Also has the property of ordering by priority following the triplet:
+/// (max_gas*gas_price, rev(emission_slot), rev(emission_index))
+pub type AsyncMessageId = (Amount, std::cmp::Reverse<Slot>, std::cmp::Reverse<u64>);
+
+/// Structure defining an asynchronous smart contract message
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct AsyncMessage {
+    /// Slot at which the message was emitted
+    pub emission_slot: Slot,
+
+    /// Index of the emitted message within the emission_slot.
+    /// This is used for disambiguating the emission of multiple messages at the same slot.
+    pub emission_index: u64,
+
+    /// The address that sent the message
+    pub sender: Address,
+
+    /// The address towards which the message is being sent
+    pub destination: Address,
+
+    /// the handler function name within the destination address' bytecode
+    pub handler: String,
+
+    /// Maximum gas to use when processing the message
+    pub max_gas: u64,
+
+    /// Gas price to take into account when executing the message.
+    /// max_gas * gas_price are burned by the sender when the emessage is sent.
+    pub gas_price: Amount,
+
+    /// Coins sent from the sender to the target address of the message.
+    /// Those coins are spent by the sender address when the emessage is sent,
+    /// and credited to the destination address when receiving the message.
+    /// In case of failure or discard, those coins are reimbursed to the sender.
+    pub coins: Amount,
+
+    /// Slot at which the message starts being valid (bound included in the validity range)
+    pub validity_start: Slot,
+
+    /// Slot at which the message stops being valid (bound not included in the validity range)
+    pub validity_end: Slot,
```

Comment:
So it's not the same validity range concept as for operations ?

Reply:
Ok so no fixed width for that range, right ? As for the upper bound not being included, is it documented somwhere for the user ?